src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -11,6 +11,28 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
+    internal enum NullableAnnotation : byte
```

Comment:
>NullableAnnotation [](start = 18, length = 18)

In an offline discussion we decided to change names of the enum elements as follows:
```
Unknown
NotAnnotated
Annotated
NotNullable
Nullable
```

Reply:
Separately from the naming, I'd still like to clarify the rules/invariants around those values.
Is it fair to say that `Unknown` and `Annotated` are persistent (we leave them alone, as we don't need to fix them with knowledge about the type), but `NotAnnotated` should always get replaced with either `NotNullable` or `Nullable` during flow analysis, at least?
If so, can we assert that the result type stored in NullableWalker never has `NotAnnotated`?

Reply:
> Is it fair to say that Unknown and Annotated are persistent (we leave them alone, as we don't need to fix them with knowledge about the type), but NotAnnotated should always get replaced with either NotNullable or Nullable during flow analysis, at least?

I am not planning to add such requirements or invariants.

---
In reply to: [231449165](https://github.com/dotnet/roslyn/pull/30913#discussion_r231449165) [](ancestors = 231449165)

Reply:
Just to make sure I understand: it is ok for the result type in NullableWalker to be an unannotated type parameter (which may or may not be possibly be null), and such a result type is not an indication that we missed a case in NullableWalker?

If that's correct, then my mental model would be:
- any time some code checks for NotAnnotated, it should also check `IsPossiblyNullableReferenceTypeTypeParameter`. And if it is post-binding, then that code should also check for NotNullable.
- conversely, any time some code checks for NotNullable, it should also check NotAnnotated+IsPossiblyNullableReferenceTypeTypeParameter

Are those rules correct?

---
In reply to: [231610532](https://github.com/dotnet/roslyn/pull/30913#discussion_r231610532) [](ancestors = 231610532,231449165)

Reply:
> Are those rules correct?

Sorry, I do not think in categories like that.

This is my view of the world:
- Any annotation can come from any code as long it reflect expected semantics
- Any time a decision has to be made whether something annotated by NotAnnotated can/can't be null, annotation and the type should be looked at together. 



---
In reply to: [231635710](https://github.com/dotnet/roslyn/pull/30913#discussion_r231635710) [](ancestors = 231635710,231610532,231449165)

Reply:
This set of nullable states contradicts the recent LDM-approved specification for handling type inference in https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-reference-types.md